src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -33,27 +34,51 @@ object RuleUtils {
    * @return indexes built for this plan
    */
   def getCandidateIndexes(indexManager: IndexManager, plan: LogicalPlan): Seq[IndexLogEntry] = {
-    // Map of a signature provider to a signature generated for the given plan.
-    val signatureMap = mutable.Map[String, String]()
-
-    def signatureValid(entry: IndexLogEntry): Boolean = {
-      val sourcePlanSignatures = entry.source.plan.properties.fingerprint.properties.signatures
-      assert(sourcePlanSignatures.length == 1)
-      val sourcePlanSignature = sourcePlanSignatures.head
-
-      if (!signatureMap.contains(sourcePlanSignature.provider)) {
-        val signature = LogicalPlanSignatureProvider
-          .create(sourcePlanSignature.provider)
-          .signature(plan)
-        signatureMap.put(sourcePlanSignature.provider, signature)
-      }
-      signatureMap(sourcePlanSignature.provider).equals(sourcePlanSignature.value)
-    }
+    getLogicalRelation(plan) match {
+      case Some(r) =>
+        // Map of a signature provider to a signature generated for the given plan.
+        val signatureMap = mutable.Map[String, String]()
+
+        def signatureValid(entry: IndexLogEntry): Boolean = {
+          val sourcePlanSignatures =
+            entry.source.plan.properties.fingerprint.properties.signatures
+          assert(sourcePlanSignatures.length == 1)
+          val sourcePlanSignature = sourcePlanSignatures.head
+
+          if (!signatureMap.contains(sourcePlanSignature.provider)) {
+            val signature = LogicalPlanSignatureProvider
+              .create(sourcePlanSignature.provider)
+              .signature(r)
+              .get // It is safe to call Option.get as signature() is
+            // called on a LogicalRelation node which is
+            // valid for signature computation
+            signatureMap.put(sourcePlanSignature.provider, signature)
+          }
 
-    // TODO: the following check only considers indexes in ACTIVE state for usage. Update
-    //  the code to support indexes in transitioning states as well.
-    val allIndexes = indexManager.getIndexes(Seq(Constants.States.ACTIVE))
+          signatureMap(sourcePlanSignature.provider).equals(sourcePlanSignature.value)
+        }
 
-    allIndexes.filter(index => index.created && signatureValid(index))
+        // TODO: the following check only considers indexes in ACTIVE state for usage. Update
+        //  the code to support indexes in transitioning states as well.
+        val allIndexes = indexManager.getIndexes(Seq(Constants.States.ACTIVE))
+        allIndexes.filter(index => index.created && signatureValid(index))
+
+      case None => Seq[IndexLogEntry]()
+    }
+  }
+
+  /**
+   * Extract LogicalRelation node from a given linear logical plan.
```

Comment:
We are not "given` a linear plan, right? You meant "Extract a LogicalRelation node if the given logical plan is linear"?

Reply:
true, I fixed it.